### PR TITLE
Add a Shutdown method to sdk TraceProvider

### LINKED ISSUE #1206: It should be easy to Shutdown() SpanProcessors
Forked from https://github.com/open-telemetry/opentelemetry-go/issues/1197#issuecomment-697132450.

[SpanProcessor](https://github.com/open-telemetry/opentelemetry-go/blob/master/sdk/trace/span_processor.go#L24) supports a Shutdown() method, which users should call if their program exits to ensure spans are written before shutting down.  Currently, many of our convenience functions, including [WithBatcher](https://github.com/open-telemetry/opentelemetry-go/blob/8c3cc4399245deab10a089024467467f450fe0d4/sdk/trace/provider.go#L193), the stdout [NewExportPipeline](https://github.com/open-telemetry/opentelemetry-go/blob/master/exporters/stdout/exporter.go#L53) and [InstallNewPipeline](https://github.com/open-telemetry/opentelemetry-go/blob/master/exporters/stdout/exporter.go#L85) don't return anything on which shutdown can be called.

One [proposed](https://github.com/open-telemetry/opentelemetry-go/issues/1197#issuecomment-698543704) solution is to add a Shutdown() method to `apitrace.TracerProvider`, which would invoke Shutdown() on all of the span processors.

cc @MrAlias @Aneurysm9

### PR DESCRIPTION
**Pull request by huikang**
- sdktraceprovider shutdown span processors
- In examples, replace processosr shutdown with
  traceprovider's shutdown

Resolves #1206 

Signed-off-by: Hui Kang <kangh@us.ibm.com>

### CODE REVIEW

File: sdk/trace/provider.go
Comment (Aneurysm9) [inline sdk/trace/provider.go:185]: Please add a docblock and tests for this method.

Comment (huikang) [reply sdk/trace/provider.go:186]: Added

Comment (Aneurysm9) [reply sdk/trace/provider.go:186]: It looks like most of the `TracerProvider` methods are covered incidentally in other tests.  This method would require a test of its own, so creating a new `provider_test.go` file and adding the test there would probably be appropriate.

File: api/trace/api.go
Comment (MrAlias) [inline api/trace/api.go:36]: I would like to get resolution on https://github.com/open-telemetry/opentelemetry-specification/issues/1073 before we merge this change to the API. Looking at other implementations this is not something they included in the API (https://github.com/open-telemetry/opentelemetry-java/blob/78fd4d1c8f4c4f85282b124db92401eb1feb28df/api/src/main/java/io/opentelemetry/trace/TracerProvider.java#L20-L43, https://github.com/open-telemetry/opentelemetry-python/blob/0e852ea393aa042d429657b9d22208eb33b52d67/opentelemetry-api/src/opentelemetry/trace/__init__.py#L172-L178, https://github.com/open-telemetry/opentelemetry-js/blob/ae1185dbcdc41d666f7e9f47474ea8996dbc8968/packages/opentelemetry-api/src/trace/tracer_provider.ts#L22-L35), rather just the SDK (https://github.com/open-telemetry/opentelemetry-java/blob/78fd4d1c8f4c4f85282b124db92401eb1feb28df/sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/TracerSdkManagement.java#L58, https://github.com/open-telemetry/opentelemetry-python/blob/0e852ea393aa042d429657b9d22208eb33b52d67/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L833, https://github.com/open-telemetry/opentelemetry-js/blob/ae1185dbcdc41d666f7e9f47474ea8996dbc8968/packages/opentelemetry-tracing/src/BasicTracerProvider.ts#L103-L105).

If we want to reduce the scope to just the SDK and unblock this from the specification issue we could follow the same by only implementing at the SDK level here.

File: sdk/trace/provider.go
Comment (MrAlias) [inline sdk/trace/provider.go:186]: Can we change the function signature to require a context be passed and return an error?

i.e.

```suggestion
func (p *TracerProvider) Shutdown(ctx context.Context) error {
```

This would mirror other `Shutdown` methods we have defined (https://github.com/open-telemetry/opentelemetry-go/blob/5660b0b59f4c0fbb9e2879bb645b632b50f5f3bd/sdk/export/trace/trace.go#L47). Additionally, it would allow for the caller to pass a deadline in the context passed, something useful to have so this doesn't block forever and can switch to a more aggressive resource recovery strategy if the deadline is exceeded. Also, the returned error would be useful in communicating failure back up the call chain if one occurs. Otherwise, downstream errors will be dropped or have to use the `ErrorHandler` interface which is less desirable than having the caller make the decision of what to do with an error.

**Review by MrAlias — CHANGES_REQUESTED** [review]
I think we can scope this to just the SDK for now and get this merged. Otherwise, we should probably wait for resolution on https://github.com/open-telemetry/opentelemetry-specification/issues/1073

File: api/trace/api.go
Comment (XSAM) [reply api/trace/api.go:36]: The solution from https://github.com/open-telemetry/opentelemetry-specification/issues/1073#issuecomment-707442967, we should add `Shutdown` to `TracerProvider` SDK but leave the `TracerProvider` interface unchanged.

Comment (huikang) [reply api/trace/api.go:36]: @XSAM , thanks for the update. If I understand you correctly, the PR shall be updated:

1. Remove the `Shutdown`  method in  the `TracerProvider interface {` in the `./trace.go`
2. Add the `shutdown` method to `TracerProvider struct ` in `./sdk/trace/provider.go`

Is my understanding correct? Thanks.

File: sdk/trace/provider.go
Comment (huikang) [reply sdk/trace/provider.go:184]: @Aneurysm9 I added `provider_test.go`. Please help to take a look. Thanks.

File: sdk/trace/provider.go
Comment (huikang) [reply sdk/trace/provider.go]: Hi, @MrAlias , I updated the signature of Shutdown to include the context in the parameter and return an error. I am not sure if it is necessary to implement checking the deadline is exceed in the method. Could you provide any guidance? Thanks.

**Review by Aneurysm9 — APPROVED** [review]

**Review by XSAM — APPROVED** [review]

File: CHANGELOG.md
Comment (MrAlias) [inline CHANGELOG.md:54]:
```suggestion
- Add Shutdown method to sdk/trace/provider, shutdown processors in the order they were registered. (#1227)
```

**Review by MrAlias — APPROVED** [review]

**Pull request merged**
